### PR TITLE
Online expansion support

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#1674

### CODE REVIEW

**Review by innobead — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
Bypassing the code factor check on purpose, but we need to revisit our code style in the future.

File: controller/engine_controller.go [hunk 1]
```diff
@@ -911,18 +911,6 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
...
-			for name, mode := range engine.Status.ReplicaModeMap {
-				if mode == longhorn.ReplicaModeWO {
-					engine.Status.ReplicaModeMap[name] = longhorn.ReplicaModeERR
-					hasRebuildingReplicas = true
-				}
-			}
-			if hasRebuildingReplicas {
-				m.logger.Warn("Cannot contain any rebuilding replicas during engine expansion, will mark all rebuilding replicas as ERROR")
-				return nil
-			}
```
Comment (PhanLe1010) [inline controller/engine_controller.go:925]: Why can we remove this logic?

**Review by PhanLe1010 — COMMENTED** [review]
LGTM 

left a small comment

File: controller/engine_controller.go (hunk 1 above)
Comment (shuo-wu) [reply controller/engine_controller.go:925]: Previously the expansion is offline and should be fast, hence waiting for the rebuilding complete is not user-friendly. Now we can have the online expansion, which does not care if rebuilding exists or not. Maybe we need to continue disabling the rebuilding for offline expansion. But we can do it after your volume controller refactoring merge.

Comment (PhanLe1010) [reply controller/engine_controller.go:925]: I see. 

But will there be a problem if we try to expand a replica when it is being rebuilt?

Comment (shuo-wu) [reply controller/engine_controller.go:925]: Based on the integration test results, No.
For the replicas, the expansion is like snapshotting with some extra operations. If operation snapshotting works fine with rebuilding, we don't need to worry about the expansion part.

Comment (PhanLe1010) [reply controller/engine_controller.go:925]: Sounds good. Thank you for the confirmation

**Pull request merged**